File: algorithms/S/Sum of Nodes with Even-Valued Grandparent/Sum of Nodes with Even-Valued Grandparent.js
const dfs = function(node, evenParent) {
    if (!node) return 0;
    
    const isEvenNode = node.val % 2 === 0;
    
    const left = dfs(node.left, isEvenNode);
    const right = dfs(node.right, isEvenNode);
    
    if (!evenParent) return left + right;
    return left + right + (node.left ? node.left.val : 0) + (node.right ? node.right.val : 0);
}

var sumEvenGrandparent = function(root) {
    if (!root) return 0;
    return dfs(root, false);
};

File: scripts/algorithms/S/Sum of Nodes with Even-Valued Grandparent/Sum of Nodes with Even-Valued Grandparent.cpp
// Runtime: 47 ms (Top 96.11%) | Memory: 41.4 MB (Top 80.98%)
class Solution {
public:
    void sumGparent(TreeNode* child, int& sum, TreeNode* parent, TreeNode* Gparent){
        if(!child) return;
        if(Gparent) if(Gparent->val % 2 ==0) sum += child->val;
        sumGparent(child->left, sum, child, parent);
        sumGparent(child->right, sum, child, parent);
    }

    int sumEvenGrandparent(TreeNode* root) {
        if(!root) return 0;
        int sum=0;
        sumGparent(root, sum, NULL, NULL); // (child, sum, parent, grand-parent)
        return sum;
    }
};